clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala

```diff
@@ -22,293 +20,191 @@ import java.time.ZoneOffset
+   *  client will similarly assume that role.
+   *
+   *  @param hc         (partial) Hadoop configuration of fs.s3a.
+   *  @param bucket     that this client will access.
+   *  @param region     to find this bucket.
+   *  @param numRetries number of times to retry on AWS.
    */
-  def leftAntiJoinAddresses(
-      leftRangeIDs: Set[String],
-      rightRangeIDs: Set[String],
-      apiConf: APIConfigurations,
-      repo: String,
-      hcValues: Broadcast[ConfMap]
-  ): Set[(String, String, Boolean, Long)] = {
-    val leftTuples = distinctEntryTuples(leftRangeIDs, apiConf, repo, hcValues)
-    val rightTuples = distinctEntryTuples(rightRangeIDs, apiConf, repo, hcValues)
-    leftTuples -- rightTuples
+  def build(hc: Configuration, bucket: String, region: String, numRetries: Int): AmazonS3
+}
+
+class GarbageCollector(val rangeGetter: RangeGetter, configMap: ConfigMapper) extends Serializable {
+  @transient lazy val spark = SparkSession.active
+
+  def getCommitsDF(runID: String, commitDFLocation: String): Dataset[Row] = {
+    spark.read
+      .option("header", value = true)
+      .option("inferSchema", value = true)
+      .csv(commitDFLocation)
   }
 
-  private def distinctEntryTuples(
-      rangeIDs: Set[String],
-      apiConf: APIConfigurations,
-      repo: String,
-      hcValues: Broadcast[ConfMap]
-  ) = {
-    // Process rangeIDs using mutation to ensure complete control over when
-    // each range is read.
-    var tuples = collection.mutable.Set[(String, String, Boolean, Long)]()
-    rangeIDs.foreach((rangeID: String) =>
-      tuples ++= getEntryTuples(rangeID, apiConf, repo, hcValues)
-    )
-    tuples.toSet
+  def getRangeIDsForCommits(commitIDs: Dataset[String], repo: String): Dataset[String] = {
+    import spark.implicits._
+
+    commitIDs.flatMap(rangeGetter.getRangeIDs(_, repo))
   }
 
-  /** receives a dataframe containing active and expired ranges and returns entries contained only in expired ranges
+  /** Compute a distinct antijoin using  partitioner.  This is the same as
+   *
+   *  <pre>
+   *     left.distinct.except(right.distinct)
+   *  </pre>
    *
-   *  @param ranges dataframe of type   rangeID:String | expired: Boolean
-   *  @return dataframe of type  key:String | address:String | relative:Boolean | last_modified:Long
+   *  but forces our plan.  Useful when Spark is unable to predict the
+   *  structure and/or size of either argument.
+   *
+   *  @return all elements in left that are not in right.
    */
-  def getExpiredEntriesFromRanges(
-      ranges: Dataset[Row],
-      apiConf: APIConfigurations,
+  def minus(
+      left: Dataset[String],
+      right: Dataset[String],
+      partitioner: Partitioner
+  ): Dataset[String] = {
+    import spark.implicits._
+    def mark(f: Int) = (p: Any) => p match { case (t: String) => (t, f) }
+    val both = left.map(mark(1)).union(right.map(mark(2)))
+    val reduced = both.rdd.reduceByKey(partitioner, _ | _)
```

Comment:
for my understanding, what is the reducer function? I can't understand by reading it. 
I expect it to generate 1 for items that are only part of left, but what would it generate for the rest?

Reply:
`_ | _` is Scala for "the bitwise OR function".  So it returns 1 for keys which have only been seen on the left, 2 for keys which have only been seen on the right, and 3 (== 1|2) for keys which have been seen on both sides.  On the next action we filter to keep only 1's -- keys which have only been seen on the left.  What can I say, this is a great time for [bit twiddling](http://catb.org/jargon/html/B/bit-twiddling.html).

Adding a name for `_ | _`, which I hope will make things understandable.